File: DynamicProgramming/21combinationSum/CombinationSum.java
public class CombinationSum {
    private static void combine(int i, int total, int target, int numsLen, int[] candidates, List<List<Integer>> res,
            List<Integer> dp) {
        if (total == target) {
            res.add(List.copyOf(dp));
            return;
        }
        if (i == numsLen) {
            return;
        }
        if (total + candidates[i] <= target) {
            total += candidates[i];
            dp.add(candidates[i]);
            combine(i, total, target, numsLen, candidates, res, dp);
            int toMin = dp.get(dp.size() - 1);
            dp.remove(dp.size() - 1);
            total -= toMin;
        }
        combine(i + 1, total, target, numsLen, candidates, res, dp);
    }

    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        // System.out.println(res);
        List<Integer> dp = new ArrayList<Integer>();
        combine(0, 0, target, candidates.length, candidates, res, dp);
        return res;
    }

    public static void main(String[] args) {
        int[] nums = { 2, 3, 5 };
        System.out.println(combinationSum(nums, 8));
    }
}

File: DynamicProgramming/21combinationSum/combinationSum.py
class Solution:
    def combine(self, nums: list[int], target: int) -> list[list[int]]:
        dp = []
        res = []
        def backtrack(i, total):
            if total == target:
                if dp in res:
                    return
                res.append(dp.copy())
                return
            if i == len(nums):
                return
            if total + nums[i] <= target:
                total += nums[i]
                dp.append(nums[i])
                backtrack(i, total)
                tomin = dp.pop()
                total -= tomin
            backtrack(i+1, total)
        backtrack(0,0)
        return res
print(Solution().combine([2,3,5], 8))
print(Solution().combine([2,3,6,7], 7))
print(Solution().combine([8,7,4,3], 11))
print(Solution().combine([2,5,3,6], 10))
print(Solution().combine([5,7,5,6,5,5,5,6,7], 5))
print(Solution().combine([1,2,3], 4))